File: examples/ios/iosApp/di/AppGraph.swift
//
//  AppGraph.swift
//  iosApp
//
//  Created by Saud Khan on 2023-09-23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import Seahorse

class AppGraph {
    static let shared: AppGraph = AppGraph()

    let stringKeys: [String]
    let seahorse: Seahorse

    private init() {
        let stringKeys = [
            "hello",
            "world",
            "platform",
            "sentence_structure",
            "foundation",
            "dune",
            "three_body_problem",
        ]
        self.stringKeys = stringKeys
        self.seahorse = Seahorse { b in
            b.fallbackSource = NSLocalizedFallbackSource()
            b.localStore = NativeSqliteLocalStore()
            b.networkSource = DarwinKtorNetworkSource(getUrlForLanguageId: { languageId in
                "https://www.bidyut.com/tech/seahorse/sample/\(languageId.lowercased()).json"
            })
        }
    }
}

File: examples/jvm_cli/src/main/kotlin/com/bidyut/tech/seahorse/example/di/AppGraph.kt
package com.bidyut.tech.seahorse.example.di

import com.bidyut.tech.seahorse.Seahorse
import com.bidyut.tech.seahorse.data.CioKtorNetworkSource
import com.bidyut.tech.seahorse.data.JdbcSqliteLocalStore
import com.bidyut.tech.seahorse.example.data.JsonResourceFallbackSource

class AppGraph {
    val stringKeys = listOf(
        "hello",
        "world",
        "platform",
        "sentence_structure",
        "foundation",
        "dune",
        "three_body_problem",
    )

    val seahorse: Seahorse by lazy {
        Seahorse {
            fallbackSource = JsonResourceFallbackSource()
            localStore = JdbcSqliteLocalStore()
            networkSource = CioKtorNetworkSource { languageId ->
                "https://www.bidyut.com/tech/seahorse/sample/${languageId.lowercase()}.json"
            }
        }
    }

    companion object {
        lateinit var instance: AppGraph
            private set

        val isAssigned: Boolean
            get() = ::instance.isInitialized

        fun assign(
            graph: AppGraph,
        ) {
            instance = graph
        }
    }
}
